Check whether locals are too large instead of whether accesses into them are too large

oli-obk opened the pull request:
Essentially this stops const prop from attempting to optimize

```rust
let mut x = [0_u8; 5000];
x[42] = 3;
```

I don't expect this to be a perf improvement without #73656 (which is also where the lack of this PR will be a perf regression).

r? @wesleywiser

oli-obk committed:
Check whether locals are too large instead of whether accesses into them are too large

oli-obk commented:
@bors try @rust-timer queue

but even if I don't expect any perf changes, let's make sure

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit b94da2bf95bc827b7f4aa2e1828fcf1aa132a3f4 with merge eb3733b311164da35f5f9e82c96c841860ee9f66...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b94da2bf95bc827b7f4aa2e1828fcf1aa132a3f4","merge_sha":"eb3733b311164da35f5f9e82c96c841860ee9f66"} -->

oli-obk committed:
Stop propagating to locals that were marks as unpropagatable.

We used to erase these values immediately after propagation, but some
things slipped through and it caused us to still initialize huge locals.

oli-obk commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit a9725ee777812a379d62de197d7c1bbed4cf45eb with merge bd190279c18106147d12a444e6ae7bc861b2de24...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a9725ee777812a379d62de197d7c1bbed4cf45eb","merge_sha":"bd190279c18106147d12a444e6ae7bc861b2de24"} -->

oli-obk committed:
Update ui tests

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/914866141), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35616)
Build commit: bd190279c18106147d12a444e6ae7bc861b2de24 (`bd190279c18106147d12a444e6ae7bc861b2de24`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35616","checks-actions":"https://github.com/rust-lang-ci/rust/runs/914866141"},"merge_sha":"bd190279c18106147d12a444e6ae7bc861b2de24"} -->

rust-timer commented:
Queued bd190279c18106147d12a444e6ae7bc861b2de24 with parent 52d2c7ac948d6abdc18eb9e05a53d03ddcaffd98, future [comparison URL](https://perf.rust-lang.org/compare.html?start=52d2c7ac948d6abdc18eb9e05a53d03ddcaffd98&end=bd190279c18106147d12a444e6ae7bc861b2de24).

rust-timer commented:
Finished benchmarking try commit (bd190279c18106147d12a444e6ae7bc861b2de24): [comparison url](https://perf.rust-lang.org/compare.html?start=52d2c7ac948d6abdc18eb9e05a53d03ddcaffd98&end=bd190279c18106147d12a444e6ae7bc861b2de24).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

wesleywiser reviewed (approved):
LGTM! Can we add a mir-opt test showing that 

```rust
let mut x = [0_u8; 5000];
x[42] = 3;
```

does not propagate? r=me with a test.

oli-obk committed:
Add test ensuring that we don't propagate large arrays

oli-obk commented:
@bors r=wesleywiser

bors commented:
:pushpin: Commit 590885ade8a4363e1a7a75a0ffeed98236df8aa5 has been approved by `wesleywiser`

<!-- @bors r=wesleywiser 590885ade8a4363e1a7a75a0ffeed98236df8aa5 -->
<!-- homu: {"type":"Approved","sha":"590885ade8a4363e1a7a75a0ffeed98236df8aa5","approver":"wesleywiser"} -->

bors commented:
:hourglass: Testing commit 590885ade8a4363e1a7a75a0ffeed98236df8aa5 with merge c0c2bd7f44ee27f17f8fbff023d61267636fa763...
<!-- homu: {"type":"BuildStarted","head_sha":"590885ade8a4363e1a7a75a0ffeed98236df8aa5","merge_sha":"c0c2bd7f44ee27f17f8fbff023d61267636fa763"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/919038523)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/919038523","builder_name":"checks-actions"} -->

oli-obk commented:
@bors r=wesleywiser

bors commented:
:pushpin: Commit 99a85e811f45b7aa02d917c44c12702d007b9131 has been approved by `wesleywiser`

<!-- @bors r=wesleywiser 99a85e811f45b7aa02d917c44c12702d007b9131 -->
<!-- homu: {"type":"Approved","sha":"99a85e811f45b7aa02d917c44c12702d007b9131","approver":"wesleywiser"} -->

bors commented:
:hourglass: Testing commit 99a85e811f45b7aa02d917c44c12702d007b9131 with merge a8bd270e3a4c736b0f98329230996d71b057d77d...
<!-- homu: {"type":"BuildStarted","head_sha":"99a85e811f45b7aa02d917c44c12702d007b9131","merge_sha":"a8bd270e3a4c736b0f98329230996d71b057d77d"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/920051395)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/920051395","builder_name":"checks-actions"} -->

oli-obk committed:
Update clippy ui test.

The reason we do not trigger these lints anymore is that clippy sets the mir-opt-level to 0, and the recent changes subtly changed how the const propagator works.

oli-obk commented:
I had to update the output of a clippy lint. clippy builds with `-Zmir-opt-level=0`, which throws off const prop completely.

@bors r=wesleywiser

bors commented:
:pushpin: Commit 303547c66285fae22396a4ecfedb6431aba8c608 has been approved by `wesleywiser`

<!-- @bors r=wesleywiser 303547c66285fae22396a4ecfedb6431aba8c608 -->
<!-- homu: {"type":"Approved","sha":"303547c66285fae22396a4ecfedb6431aba8c608","approver":"wesleywiser"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #74837) made this pull request unmergeable. Please resolve the merge conflicts.

oli-obk committed:
Update tests after rebase

oli-obk committed:
Improve the diagnostics around misspelled mir dump filenames

oli-obk commented:
I made some changes to the mir-opt test suite. It now prints the name of the mir dump file it's expecting and the path to the file which requested the mir dump

xldenis commented on src/tools/compiletest/src/runtest.rs:
```diff
@@ -3239,8 +3239,19 @@ impl<'test> TestCx<'test> {
+        let to_full_path = |path: PathBuf| {
+            let full = self.get_mir_dump_dir().join(&path);
+            if !full.exists() {
+                panic!(
```
👍 there are a couple other places in `check_mir_dump` that also produce confusing / less useful errors but that could be fixed in a seperate pr.

wesleywiser commented:
@bors r+

bors commented:
:pushpin: Commit 1864a973b3a0a5a6c5a7c71d7d7cd052732e5c02 has been approved by `wesleywiser`

<!-- @bors r=wesleywiser 1864a973b3a0a5a6c5a7c71d7d7cd052732e5c02 -->
<!-- homu: {"type":"Approved","sha":"1864a973b3a0a5a6c5a7c71d7d7cd052732e5c02","approver":"wesleywiser"} -->

bors commented:
:hourglass: Testing commit 1864a973b3a0a5a6c5a7c71d7d7cd052732e5c02 with merge 4d4342347b71313258a46e506ee0a258f365185c...
<!-- homu: {"type":"BuildStarted","head_sha":"1864a973b3a0a5a6c5a7c71d7d7cd052732e5c02","merge_sha":"4d4342347b71313258a46e506ee0a258f365185c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/959133317), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35851)
Approved by: wesleywiser
Pushing 4d4342347b71313258a46e506ee0a258f365185c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"wesleywiser","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35851","checks-actions":"https://github.com/rust-lang-ci/rust/runs/959133317"},"merge_sha":"4d4342347b71313258a46e506ee0a258f365185c"} -->
